src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,113 @@ impl<S: StateStore> StateTable<S> {
-        _stats: Arc<StateStoreMetrics>,
-    ) -> StorageResult<Self> {
-        todo!()
+        keyspace: Keyspace<S>,
+        table_descs: Vec<ColumnDesc>,
+        mem_table_iter: Peekable<MemTableIter<'a>>,
+        order_types: Vec<OrderType>,
+        epoch: u64,
+    ) -> StorageResult<StateTableRowIter<'a, S>> {
+        let cell_based_streaming_iter =
+            CellBasedTableStreamingIter::new(&keyspace, &table_descs, epoch).await?;
+        let state_table_iter = Self {
+            mem_table_iter,
+            cell_based_streaming_iter,
+            cell_based_item: None,
+            order_types,
+        };
+
+        Ok(state_table_iter)
     }
 
-    async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let cell_based_item;
+        match &self.cell_based_item {
+            Some(remained_cell_based_item) => {
+                cell_based_item = Some(remained_cell_based_item.clone());
+                self.cell_based_item = self.cell_based_streaming_iter.next().await?;
+            }
+            None => cell_based_item = self.cell_based_streaming_iter.next().await?,
+        }
+
+        let mut mem_table_iter_next_flag = false;
+        let mut res = None;
+        match (cell_based_item, self.mem_table_iter.peek()) {
+            (None, None) => {
+                return Ok(None);
+            }
+            (Some((_, row)), None) => {
+                return Ok(Some(row));
+            }
+            (None, Some((_, row_op))) => {
+                mem_table_iter_next_flag = true;
+                match row_op {
+                    RowOp::Insert(row) => res = Some(row.clone()),
+                    RowOp::Delete(_) => {}
+                    RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
+                }
+            }
+            (Some((cell_based_pk, cell_based_row)), Some((mem_table_pk, mem_table_row_op))) => {
+                let pk_serializer = OrderedRowSerializer::new(self.order_types.clone());
+                let mem_table_pk_bytes = serialize_pk(mem_table_pk, &pk_serializer).map_err(err)?;
+                match cell_based_pk.cmp(&mem_table_pk_bytes) {
+                    Ordering::Less => {
+                        res = Some(cell_based_row);
+                    }
+                    Ordering::Equal => {
+                        mem_table_iter_next_flag = true;
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => {
+                                res = Some(row.clone());
+                            }
+                            RowOp::Delete(_) => {}
+                            RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
+                        }
+                    }
+                    Ordering::Greater => {
+                        // mem_table_item will be return, while cell_based_item need to be saved and
```

Comment:
incorrect. what if the next mem table item is still greater than cell based item?

Reply:
I'm thinking of if we can use async_stream to implement this part. Otherwise it's too hard to understand and prune to error.

Reply:
please revisit and update the comments